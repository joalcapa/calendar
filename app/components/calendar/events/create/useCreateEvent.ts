'use client'

import { useState, useMemo } from 'react';
import { Event } from '@/types/event';
import useGetEvent from '@/app/hooks/useGetEvent';

interface TargetEvent {
  target: {
    value: string,
  },
}

interface TargetCheckEvent {
  target: {
    checked: boolean,
  },
}

interface CreateEventProps {
  onClose: () => void;
  event?: Event,
  isDelete?: boolean;
}

const formatDateForInput = (dateString: string) => {
  const date = new Date(dateString);
  const year = date.getFullYear();
  const month = String(date.getMonth() + 1).padStart(2, '0');
  const day = String(date.getDate()).padStart(2, '0');
  const hours = String(date.getHours()).padStart(2, '0');
  const minutes = String(date.getMinutes()).padStart(2, '0');
  return `${year}-${month}-${day}T${hours}:${minutes}`;
};

const useCreateEvent = (props: CreateEventProps) => {
  const { onClose = () => { }, event, isDelete } = props;
  const [loading, setLoading] = useState(false);
  const [title, setTitle] = useState(event && event.title ? event.title : '');
  const [description, setDescription] = useState(event && event.description ? event.description : '');
  const [startDate, setStartDate] = useState(event && event.start_date ? formatDateForInput(event.start_date.toString()) : '');
  const [isAllDay, setAllDay] = useState(event && event.is_all_day ? event.is_all_day : false);
  const [finishDate, setFinishDate] = useState(event && event.finish_date ? formatDateForInput(event.finish_date.toString()) : '');
  const { isLoading, error } = useGetEvent();

  const onCreate = async () => {

  };

  const resetForm = () => {
    setTitle('');
    setDescription('');
    setStartDate('');
    setAllDay(false);
    setFinishDate('');
  };

  const changeTitle = (event: TargetEvent) => {
    setTitle(event.target.value);
  };

  const changeDescription = (event: TargetEvent) => {
    setDescription(event.target.value);
  };

  const changeStartDate = (event: TargetEvent) => {
    setStartDate(event.target.value);
  };

  const changeFinishDate = (event: TargetEvent) => {
    setFinishDate(event.target.value);
  };

  const changeAllDay = (event: TargetCheckEvent) => {
    setAllDay(event.target.checked);
  };

  const isValidForm = useMemo(() => {
    return !!title && !!description && !!startDate && (isAllDay ? true : !!finishDate);
  }, [title, description, startDate, isAllDay, finishDate, event]);

  return {
    isDelete,
    isLoading,
    error,
    title,
    description,
    startDate,
    isAllDay,
    finishDate,
    onClose,
    changeTitle,
    changeDescription,
    changeStartDate,
    changeAllDay,
    changeFinishDate,
    onCreate,
    isValidForm,
  };
};

export default useCreateEvent;