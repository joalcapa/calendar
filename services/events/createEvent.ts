import { eventRepository } from '@/repositories/eventRepository';
import { Event } from '@/types/event';
import BaseService from '../baseService';

export default class CreateEvent extends BaseService {
  private data: Event;
  private event: Event | null;

  constructor(data: Event) {
    super();
    this.data = data;
    this.event = null;
  }

  public async call(): Promise<void> {
    if (!this.data.title) {
      this.setError('El título es requerido');
      return
    }

    if (!this.data.description) {
      this.setError('La descripción es requerida');
      return
    }

    if (!this.data.start_date) {
      this.setError('La Fecha de inicio es requerida');
      return
    }

    if (!this.data.is_all_day && !this.data.finish_date) {
      this.setError('La Fecha de culminación es requerida');
      return
    }

    const startDate = new Date(this.data.start_date);
    if (this.data.is_all_day) {
      startDate.setHours(0, 0, 0, 0);
      const finishDate = new Date(startDate);
      finishDate.setUTCHours(23, 59, 59, 999);
      this.data.finish_date = finishDate;
    } else {
      const finishDate = new Date(this.data.finish_date);
      this.data.finish_date = finishDate;
    }

    this.data.start_date = startDate;

    this.event = await eventRepository.create(this.data);
  }

  public getEvent(): Event | null {
    return this.event;
  }
}