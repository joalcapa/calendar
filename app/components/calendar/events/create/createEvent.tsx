'use client'

import React from 'react';
import useCreateEvent from '../../../../hooks/useCreateEvent';
import Image from 'next/image';

interface CreateEventProps {
  isLoading: boolean;
  weather: string;
  weatherUrl: string;
  error: string;
  title: string;
  description: string;
  city: string;
  startDate: string;
  isAllDay: boolean;
  finishDate: string;
  onClose: () => void;
  changeTitle: () => void;
  changeDescription: () => void;
  changeCity: () => void;
  changeStartDate: () => void;
  changeAllDay: () => void;
  changeFinishDate: () => void;
  onSend: () => void;
  operationLabel?: string;
  buttonLabel: string;
  isValidForm: boolean;
  isDelete: boolean;
  onDelete: () => void;
  isCreating: boolean;
};

export const CreateEvent = ({
  isLoading = false,
  weather = '',
  weatherUrl = '',
  error = '',
  title = '',
  description = '',
  city = '',
  startDate = '',
  isAllDay = false,
  finishDate = '',
  changeTitle = () => { },
  changeDescription = () => { },
  changeCity = () => { },
  changeStartDate = () => { },
  changeAllDay = () => { },
  changeFinishDate = () => { },
  onSend = () => { },
  onClose = () => { },
  operationLabel = 'Crear Evento',
  buttonLabel = '¡Crear ahora!',
  isValidForm = false,
  isDelete = false,
  onDelete = () => { },
  isCreating = false,
}: CreateEventProps) => (
  <div className="fixed inset-0 flex items-center justify-center z-50 bg-black bg-opacity-50">
    <div className="bg-white p-6 rounded shadow-lg w-96">
      <div className="flex justify-between items-center mb-4">
        <div className="flex flex-row">
          <div>
            <h2 className="text-lg font-semibold">{operationLabel}</h2>
            { <div className="text-xs">{weather || ( city ? "Cargando clima..." : "Aquí podrás conocer el clima" )}</div> }
          </div>
          { !!weatherUrl && (
              <div className="pl-2">
                <Image
                    src={`https://${weatherUrl}`}
                    alt={weather}
                    width={50}
                    height={50}
                    layout="fixed"
                />
              </div>
          )}
        </div>
        {
          isDelete && (
              <button
                onClick={onDelete}
                className="text-red-500 hover:text-red-700"
                type="button"
              >
                Borrar
              </button>
            )
        }
        { ( isCreating || isLoading ) && <div className="loader ease-linear rounded-full border-4 border-t-4 border-gray-200 h-8 w-8"></div> }
      </div>
      {error && <p className="text-red-500 mb-4">{error}</p>}
      <div>
        <div className="mb-4">
          <label className="block mb-1">Título:</label>
          <input
              data-testid="title"
              type="text"
              className="border border-gray-300 p-2 w-full rounded"
              value={title}
              onChange={changeTitle}
              required
              disabled={isCreating}
          />
        </div>
        <div className="mb-4">
          <label className="block mb-1">Descripción:</label>
          <textarea
              className="border border-gray-300 p-2 w-full rounded"
              value={description}
              onChange={changeDescription}
              required
              disabled={isCreating}
          />
        </div>
        <div className="mb-4">
          <label className="block mb-1">Ciudad:</label>
          <input
              className="border border-gray-300 p-2 w-full rounded"
              value={city}
              onChange={changeCity}
              required
              disabled={isCreating}
          />
        </div>
        <div className="mb-4">
          <label className="block mb-1">Fecha de Inicio:</label>
          <input
              type="datetime-local"
              className="border border-gray-300 p-2 w-full rounded"
              value={startDate}
              onChange={changeStartDate}
              required
              disabled={isCreating}
          />
        </div>
        <div className="mb-4 flex items-center">
          <input
              type="checkbox"
              checked={isAllDay}
              onChange={changeAllDay}
              className="mr-2"
              disabled={isCreating}
          />
          <label>Todo el día</label>
        </div>
        {!isAllDay && (
            <div className="mb-4">
              <label className="block mb-1">Fecha de Finalización:</label>
              <input
                  type="datetime-local"
                  className="border border-gray-300 p-2 w-full rounded"
                  value={finishDate}
                  min={startDate}
                  onChange={changeFinishDate}
                  disabled={isCreating}
              />
            </div>
        )}
        <button
            onClick={onSend}
            disabled={isCreating || !isValidForm}
            className={`${isValidForm ? 'bg-blue-500' : 'bg-gray-500'} text-white p-2 rounded w-full`}
        >
          {buttonLabel}
        </button>
      </div>
      <button onClick={onClose} className="mt-4 text-gray-500" disabled={isCreating}>
        Cancelar
      </button>
    </div>
  </div>
);

export default (props) => {
  const hook = useCreateEvent(props);
  return <CreateEvent {...hook} />;
};
