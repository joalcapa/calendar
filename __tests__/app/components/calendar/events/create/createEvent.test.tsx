import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { CreateEvent } from '../../../../../../app/components/calendar/events/create/createEvent';

describe('CreateEvent Component', () => {
    const defaultProps = {
        isLoading: false,
        weather: '',
        weatherUrl: '',
        error: '',
        title: '',
        description: '',
        city: '',
        startDate: '',
        isAllDay: false,
        finishDate: '',
        onClose: jest.fn(),
        changeTitle: jest.fn(),
        changeDescription: jest.fn(),
        changeCity: jest.fn(),
        changeStartDate: jest.fn(),
        changeAllDay: jest.fn(),
        changeFinishDate: jest.fn(),
        onSend: jest.fn(),
        operationLabel: 'Crear Evento',
        buttonLabel: '¡Crear ahora!',
        isValidForm: false,
        isDelete: false,
        onDelete: jest.fn(),
        isCreating: false,
    };

    it('should render without crashing', () => {
        render(<CreateEvent {...defaultProps} />);
        expect(screen.getByText('Crear Evento')).toBeInTheDocument();
        expect(screen.getByText('¡Crear ahora!')).toBeInTheDocument();
    });

    it('should display the weather or default message', () => {
        render(<CreateEvent {...defaultProps} city="Bogotá" />);
        expect(screen.getByText('Cargando clima...')).toBeInTheDocument();
    });

    it('should trigger changeTitle when typing in the title input', () => {
        render(<CreateEvent {...defaultProps} />);
        const titleInput = screen.getByTestId('title');
        fireEvent.change(titleInput, { target: { value: 'Nuevo evento' } });
        expect(defaultProps.changeTitle).toHaveBeenCalled();
    });

    it('should disable inputs and button while creating', () => {
        render(<CreateEvent {...defaultProps} isCreating={true} />);
        const titleInput = screen.getAllByTestId('title');
        const sendButton = screen.getByText('¡Crear ahora!');
        expect(titleInput[0]).toBeDisabled();
        expect(sendButton).toBeDisabled();
    });

    it('should show error message if error prop is provided', () => {
        render(<CreateEvent {...defaultProps} error="Error de validación" />);
        expect(screen.getByText('Error de validación')).toBeInTheDocument();
    });

    it('should display delete button if isDelete is true', () => {
        render(<CreateEvent {...defaultProps} isDelete={true} />);
        expect(screen.getByText('Borrar')).toBeInTheDocument();
    });

    it('should trigger onSend when the button is clicked', () => {
        render(<CreateEvent {...defaultProps} isValidForm={true} />);
        const sendButton = screen.getByText('¡Crear ahora!');
        fireEvent.click(sendButton);
        expect(defaultProps.onSend).toHaveBeenCalled();
    });

    it('should trigger onClose when cancel button is clicked', () => {
        render(<CreateEvent {...defaultProps} />);
        const cancelButton = screen.getByText('Cancelar');
        fireEvent.click(cancelButton);
        expect(defaultProps.onClose).toHaveBeenCalled();
    });

    it('should not display finish date input if isAllDay is checked', () => {
        render(<CreateEvent {...defaultProps} isAllDay={true} />);
        expect(screen.queryByLabelText('Fecha de Finalización:')).not.toBeInTheDocument();
    });

    it('renders correctly when not loading', () => {
        const { asFragment } = render(<CreateEvent {...defaultProps} title="Evento de prueba" isLoading={false} onSend={jest.fn()} />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('renders correctly when loading', () => {
        const { asFragment } = render(<CreateEvent {...defaultProps} title="Evento de prueba" isLoading={true} onSend={jest.fn()} />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('renders error message when error occurs', () => {
        const { asFragment } = render(<CreateEvent {...defaultProps} title="Evento de prueba" isLoading={false} error="Error de validación" onSend={jest.fn()} />);
        expect(asFragment()).toMatchSnapshot();
    });
});
