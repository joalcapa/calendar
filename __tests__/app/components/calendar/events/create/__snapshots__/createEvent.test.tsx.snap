// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateEvent Component renders correctly when loading 1`] = `
<DocumentFragment>
  <div
    class="fixed inset-0 flex items-center justify-center z-50 bg-black bg-opacity-50"
  >
    <div
      class="bg-white p-6 rounded shadow-lg w-96"
    >
      <div
        class="flex justify-between items-center mb-4"
      >
        <div
          class="flex flex-row"
        >
          <div>
            <h2
              class="text-lg font-semibold"
            >
              Crear Evento
            </h2>
            <div
              class="text-xs"
            >
              Aquí podrás conocer el clima
            </div>
          </div>
        </div>
        <div
          class="loader ease-linear rounded-full border-4 border-t-4 border-gray-200 h-8 w-8"
        />
      </div>
      <div>
        <div
          class="mb-4"
        >
          <label
            class="block mb-1"
          >
            Título:
          </label>
          <input
            class="border border-gray-300 p-2 w-full rounded"
            data-testid="title"
            disabled=""
            required=""
            type="text"
            value="Evento de prueba"
          />
        </div>
        <div
          class="mb-4"
        >
          <label
            class="block mb-1"
          >
            Descripción:
          </label>
          <textarea
            class="border border-gray-300 p-2 w-full rounded"
            disabled=""
            required=""
          />
        </div>
        <div
          class="mb-4"
        >
          <label
            class="block mb-1"
          >
            Ciudad:
          </label>
          <input
            class="border border-gray-300 p-2 w-full rounded"
            disabled=""
            required=""
            value=""
          />
        </div>
        <div
          class="mb-4"
        >
          <label
            class="block mb-1"
          >
            Fecha de Inicio:
          </label>
          <input
            class="border border-gray-300 p-2 w-full rounded"
            disabled=""
            required=""
            type="datetime-local"
            value=""
          />
        </div>
        <div
          class="mb-4 flex items-center"
        >
          <input
            class="mr-2"
            disabled=""
            type="checkbox"
          />
          <label>
            Todo el día
          </label>
        </div>
        <div
          class="mb-4"
        >
          <label
            class="block mb-1"
          >
            Fecha de Finalización:
          </label>
          <input
            class="border border-gray-300 p-2 w-full rounded"
            disabled=""
            min=""
            type="datetime-local"
            value=""
          />
        </div>
        <button
          class="bg-gray-500 text-white p-2 rounded w-full"
          disabled=""
        >
          ¡Crear ahora!
        </button>
      </div>
      <button
        class="mt-4 text-gray-500"
        disabled=""
      >
        Cancelar
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`CreateEvent Component renders correctly when not loading 1`] = `
<DocumentFragment>
  <div
    class="fixed inset-0 flex items-center justify-center z-50 bg-black bg-opacity-50"
  >
    <div
      class="bg-white p-6 rounded shadow-lg w-96"
    >
      <div
        class="flex justify-between items-center mb-4"
      >
        <div
          class="flex flex-row"
        >
          <div>
            <h2
              class="text-lg font-semibold"
            >
              Crear Evento
            </h2>
            <div
              class="text-xs"
            >
              Aquí podrás conocer el clima
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="mb-4"
        >
          <label
            class="block mb-1"
          >
            Título:
          </label>
          <input
            class="border border-gray-300 p-2 w-full rounded"
            data-testid="title"
            required=""
            type="text"
            value="Evento de prueba"
          />
        </div>
        <div
          class="mb-4"
        >
          <label
            class="block mb-1"
          >
            Descripción:
          </label>
          <textarea
            class="border border-gray-300 p-2 w-full rounded"
            required=""
          />
        </div>
        <div
          class="mb-4"
        >
          <label
            class="block mb-1"
          >
            Ciudad:
          </label>
          <input
            class="border border-gray-300 p-2 w-full rounded"
            required=""
            value=""
          />
        </div>
        <div
          class="mb-4"
        >
          <label
            class="block mb-1"
          >
            Fecha de Inicio:
          </label>
          <input
            class="border border-gray-300 p-2 w-full rounded"
            required=""
            type="datetime-local"
            value=""
          />
        </div>
        <div
          class="mb-4 flex items-center"
        >
          <input
            class="mr-2"
            type="checkbox"
          />
          <label>
            Todo el día
          </label>
        </div>
        <div
          class="mb-4"
        >
          <label
            class="block mb-1"
          >
            Fecha de Finalización:
          </label>
          <input
            class="border border-gray-300 p-2 w-full rounded"
            min=""
            type="datetime-local"
            value=""
          />
        </div>
        <button
          class="bg-gray-500 text-white p-2 rounded w-full"
          disabled=""
        >
          ¡Crear ahora!
        </button>
      </div>
      <button
        class="mt-4 text-gray-500"
      >
        Cancelar
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`CreateEvent Component renders error message when error occurs 1`] = `
<DocumentFragment>
  <div
    class="fixed inset-0 flex items-center justify-center z-50 bg-black bg-opacity-50"
  >
    <div
      class="bg-white p-6 rounded shadow-lg w-96"
    >
      <div
        class="flex justify-between items-center mb-4"
      >
        <div
          class="flex flex-row"
        >
          <div>
            <h2
              class="text-lg font-semibold"
            >
              Crear Evento
            </h2>
            <div
              class="text-xs"
            >
              Aquí podrás conocer el clima
            </div>
          </div>
        </div>
      </div>
      <p
        class="text-red-500 mb-4"
      >
        Error de validación
      </p>
      <div>
        <div
          class="mb-4"
        >
          <label
            class="block mb-1"
          >
            Título:
          </label>
          <input
            class="border border-gray-300 p-2 w-full rounded"
            data-testid="title"
            required=""
            type="text"
            value="Evento de prueba"
          />
        </div>
        <div
          class="mb-4"
        >
          <label
            class="block mb-1"
          >
            Descripción:
          </label>
          <textarea
            class="border border-gray-300 p-2 w-full rounded"
            required=""
          />
        </div>
        <div
          class="mb-4"
        >
          <label
            class="block mb-1"
          >
            Ciudad:
          </label>
          <input
            class="border border-gray-300 p-2 w-full rounded"
            required=""
            value=""
          />
        </div>
        <div
          class="mb-4"
        >
          <label
            class="block mb-1"
          >
            Fecha de Inicio:
          </label>
          <input
            class="border border-gray-300 p-2 w-full rounded"
            required=""
            type="datetime-local"
            value=""
          />
        </div>
        <div
          class="mb-4 flex items-center"
        >
          <input
            class="mr-2"
            type="checkbox"
          />
          <label>
            Todo el día
          </label>
        </div>
        <div
          class="mb-4"
        >
          <label
            class="block mb-1"
          >
            Fecha de Finalización:
          </label>
          <input
            class="border border-gray-300 p-2 w-full rounded"
            min=""
            type="datetime-local"
            value=""
          />
        </div>
        <button
          class="bg-gray-500 text-white p-2 rounded w-full"
          disabled=""
        >
          ¡Crear ahora!
        </button>
      </div>
      <button
        class="mt-4 text-gray-500"
      >
        Cancelar
      </button>
    </div>
  </div>
</DocumentFragment>
`;
